dipy/direction/bootstrap_direction_getter.py

```diff
@@ -0,0 +1,130 @@
+                sh_order = model.sh_order
+            else:
+                sh_order = default_SH
+            
+        self.where_dwi = shm.lazy_index(~model.gtab.b0s_mask)
+        if not isinstance(self.where_dwi, slice):
+            msg = ("For optimal bootstrap tracking consider reordering the "
+                   "diffusion volumes so that all the b0 volumes are at the "
+                   "beginning")
+            warn(msg)
+        x, y, z = model.gtab.gradients[self.where_dwi].T
+        r, theta, phi = cart2sphere(x, y, z)
+        b_range = (r.max() - r.min()) / r.min()
+        if b_range > tol:
+            raise ValueError("BootOdfGen only supports single shell data")
+
+        B, m, n = shm.real_sym_sh_basis(sh_order, theta, phi)
+        H = shm.hat(B)
+        R = shm.lcr_matrix(H)
+
+        self.data = np.asarray(data, "float64")
+        self.model = model
+        self.sphere = sphere
+        self.H = H
+        self.R = R
+
+    def get_pmf(self, point):
+        """Produces an ODF from a SH bootstrap sample"""
+        single_vox_data = trilinear_interpolate4d(self.data, point)
+
+        non_b0_data = self.where_dwi
+        dwidata = single_vox_data[non_b0_data]
+        bootdata = shm.bootstrap_data_voxel(dwidata, self.H, self.R)
+        single_vox_data[non_b0_data] = bootdata
+        fit = self.model.fit(single_vox_data)
+        pmf = fit.odf(self.sphere)
+        return pmf
+
+    def pmf_no_boot(self, point):
+        data = trilinear_interpolate4d(self.data, point)
+        fit = self.model.fit(data)
+        return fit.odf(self.sphere)
+
+
+class BootDirectionGetter(BaseDirectionGetter):
+
+    def __init__(self, pmfgen, maxangle, sphere=default_sphere,
+                 max_attempts=5, **kwargs):
+        self.max_attempts = max_attempts
+        super(BootDirectionGetter, self).__init__(pmfgen, maxangle, sphere,
+                                                  **kwargs)
+
+    @classmethod
+    def from_data(cls, data, model, max_angle, sphere=default_sphere,
+                  sh_order=None, max_attempts=5, **kwargs):
+        """Create a BootDirectionGetter using HARDI data and an ODF type model
+        
+        Parameters
+        ----------
+        data : ndarray, float, (..., N)
+            Diffusion MRI data with N volumes.
+        model : dipy diffusion model
+            Must provide fit with odf method.
+        max_angle : float (0, 90)
+            Maximum angle between tract segments. This angle can be more
+            generous (larger) than values typically used with probabilistic
+            direction getters.
+        sphere : Sphere
+            The sphere used to sample the diffusion ODF.
+        sh_order : even int
+            The order of the SH "model" used to estimate bootstrap residuals.
+        max_attempts : int
+            Max number of bootstrap samples used to find tracking direction
+            before giving up.
+        pmf_threshold : float
```

Comment:
unexpected parameter in docstring. It should be removed or added as function parameter

Reply:
These get consumed by the constructor of the parent class, I'd leave them in the docstring for clarity.

Reply:
ok, Thanks for the clarification